hax/hax/server.py

```diff
@@ -259,6 +262,7 @@ def run(
             self._start(web_address, port)
             LOG.debug('Server stopped normally')
         finally:
+            self.hax_state.set_stopping()
```

Comment:
Okay, to understand correctly, hax is stopping but `repeat_if_fails()` loop is blocking it. But if finally is invoked which means  all the other threads have stopped, in which thread's context will the consul call be running? Wouldn't it be that hax never reach this statement because if `repeat_if_fails()`?

Reply:
No, this is the main thread that serves the aiohttp server only (`self._start` above is a blocking call, it exits only when the main thread has received a signal). 

By default aiohttp's `run_app` handles process signals and exits when SIGTERM is received (see `handle_signals=True` in the [doc](https://docs.aiohttp.org/en/stable/web_reference.html#utilities)).
So when we're in `finally` block, a signal is already received and understood by the web server. The only thing left is to notify all our custom threads that we're shutting down.

Those threads can potentially communicate to Consul at the moment. But hence this code has modified the global `stopping` flag in HaxGlobalState, none of Consul errors would cause those threads to hang instead of exiting.

Reply:
Ah, you are right, its the server.py not hax.py. Got it!